### DISCUSSION TITLE
PHPBench API

### DISCUSSION THREAD

**Discussion by theofidry** [phpbench/phpbench#1044] [General]
I will link my PR as soon as it's ready but I have something along those lines at the moment:

```php
final class CompileBench
{
    private const WITHOUT_COMPACTORS_DIR = __DIR__.'/../../fixtures/bench/without-compactors';
    private const WITH_COMPACTORS_DIR = __DIR__.'/../../fixtures/bench/with-compactors';

    private ApplicationRunner $runner;

    public function __construct()
    {
        $this->runner = new ApplicationRunner(
            new Application(
                autoExit: false,
                catchExceptions: false,
            ),
        );
    }

    #[Warmup(1)]
    public function warmUp(): void
    {
        FS::remove(
            __DIR__.'/../../dist/bench/box.phar',
        );

        self::removeOutputArtifact();
        self::assertVendorsAreInstalled();
    }

    private static function removeOutputArtifact(): void
    {
        FS::remove(__DIR__.'/../../dist/bench/box.phar');
    }

    private static function assertVendorsAreInstalled(): void
    {
        $vendorDirs = [
            self::WITH_COMPACTORS_DIR.'/vendor',
            self::WITHOUT_COMPACTORS_DIR.'/vendor',
        ];

        foreach ($vendorDirs as $vendorDir) {
            Assert::assertDirectoryExists($vendorDir);
        }
    }

    #[ParamProviders('parameterProvider')]
    #[Iterations(10)]
    public function bench(array $params): void
    {
        [$workingDirectory, $enableParallelization] = $params;

        chdir($workingDirectory);

        $exitCode = $this->runner->run(
            self::createIO($enableParallelization),
        );

        Assert::assertSame(ExitCode::SUCCESS, $exitCode);
    }

    public static function parameterProvider(): iterable
    {
        yield [
            self::WITHOUT_COMPACTORS_DIR,
            false,
        ];

        yield [
            self::WITH_COMPACTORS_DIR,
            false,
        ];

        yield [
            self::WITH_COMPACTORS_DIR,
            true,
        ];
    }

    private static function createIO(bool $enableParallelization): IO
    {
        $input = [
            'compile',
        ];

        if (!$enableParallelization) {
            $input['--no-parallel'] = null;
        }

        return new IO(
            new ArrayInput($input),
            new NullOutput(),
        );
    }
}

```

My first impressions:

- I really like the tool 😄 
- I am a bit surprised that `Warmup` requires a parameter, at least in this case I do not see the need of having more than 1 revolution and I wonder if there is really many cases where you would want more than one.
- A tad annoying that the param provider parameters are not extended to the method like PHPUnit does.
- There is one irk with the current API to define what you need to benchmark*.

*: I want to expend on this point in this discussion. As you can see above I have:

- A couple of parameters to configure the application execution (which is what is being tested).
- Like I assume a number of benchmarking, there is a warmup step, unfortunately I do not have access to the input parameters during this warmup phase so the `chdir()` call need to be done in the `::bench()` instead of the `::warmUp()`.
- Just to make sure I am not benchmarking something silly, I am tested the exit code (in this case I need no more than this). But to me it feels wrong that this is part of the `::bench()` as I do not want to be benchmarked.
- The point above makes me wonder if it would not be better to allow an assertion API within the test (not via attributes). Like I am testing the exit it would feel natural to test the benchmark metrics output.
- There is no tearDown counterpart to `Warmup()`. I guess I can do this in `__destruct()`. I admit I don't like that much but might be more a me problem 😅 

So I am wondering if it would not be possible to do something like this:

```php
<?php

declare(strict_types=1);

namespace KevinGH\Box\Benchmark;

use Fidry\Console\Application\ApplicationRunner;
use Fidry\Console\ExitCode;
use Fidry\Console\IO;
use Fidry\FileSystem\FS;
use KevinGH\Box\Console\Application;
use PhpBench\Attributes\Iterations;
use PhpBench\Attributes\ParamProviders;
use PhpBench\Attributes\Revs;
use PhpBench\Attributes\Warmup;
use PHPUnit\Framework\Assert;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Input\StringInput;
use Symfony\Component\Console\Output\NullOutput;
use function Safe\chdir;

final class CompileBenchDemo
{
    private const WITHOUT_COMPACTORS_DIR = __DIR__.'/../../fixtures/bench/without-compactors';
    private const WITH_COMPACTORS_DIR = __DIR__.'/../../fixtures/bench/with-compactors';

    private ApplicationRunner $runner;

    public function __construct()
    {
        $this->runner = new ApplicationRunner(
            new Application(
                autoExit: false,
                catchExceptions: false,
            ),
        );
    }

    private function warmUp(string $workingDirectory): void
    {
        FS::remove(
            __DIR__.'/../../dist/bench/box.phar',
        );

        self::removeOutputArtifact();
        self::assertVendorsAreInstalled();

        chdir($workingDirectory);
    }

    private static function removeOutputArtifact(): void
    {
        FS::remove(__DIR__.'/../../dist/bench/box.phar');
    }

    private static function assertVendorsAreInstalled(): void
    {
        $vendorDirs = [
            self::WITH_COMPACTORS_DIR.'/vendor',
            self::WITHOUT_COMPACTORS_DIR.'/vendor',
        ];

        foreach ($vendorDirs as $vendorDir) {
            Assert::assertDirectoryExists($vendorDir);
        }
    }

    #[ParamProviders('parameterProvider')]
    #[Iterations(10)]
    public function __invoke(
        Bencher $bencher,
        string $workingDirectory,
        bool $enableParallelization,
    ): void
    {
        $this->warmUp($workingDirectory);

        $result = $bencher->bench(
            fn () => $this->bench($enableParallelization),
        );

        $exitCode = $result->output;
        Assert::assertSame(ExitCode::SUCCESS, $exitCode);

        $result
            ->assertMode('variant.time.avg', '< 200 microseconds +/- 10%')
            ->assertMode('variant.time.net', '< 10');
    }

    private function bench(bool $enableParallelization): int
    {
        return $this->runner->run(
            self::createIO($enableParallelization),
        );
    }

    public static function parameterProvider(): iterable
    {
        yield [
            self::WITHOUT_COMPACTORS_DIR,
            false,
        ];

        yield [
            self::WITH_COMPACTORS_DIR,
            false,
        ];

        yield [
            self::WITH_COMPACTORS_DIR,
            true,
        ];
    }

    private static function createIO(bool $enableParallelization): IO
    {
        $input = [
            'compile',
        ];

        if (!$enableParallelization) {
            $input['--no-parallel'] = null;
        }

        return new IO(
            new ArrayInput($input),
            new NullOutput(),
        );
    }
}
```

The key idea being you would have a `Bencher` class which defines what is benchmarked exactly. You remove the need of many hooks like the warmup or other as the user has complete freedom of this. And the result could contain both the output of the benchmarked closure and also provide the benchmark results which you could do assertion againsts.

WDYT?

**Comment by dantleech** [comment]
wow, lots to unpack there :slightly_smiling_face: 

> I am a bit surprised that Warmup requires a parameter, at least in this case I do not see the need of having more than 1 revolution
it would make sense to default at 1 and happy for that to be implemented :+1: I'm not sure if there would ever be a case that requires > 1, I guess it's a valid requirement somewhere 

 > A tad annoying that the param provider parameters are not extended to the method like PHPUnit does.

Yes -- it is annoying, it seems it was for a reason though: https://github.com/phpbench/phpbench/issues/190 - TL;DR; it increases the overhead of the method call depending on the number arguments, but while that is negligible (it's constant at least) more importantly it coupled the before/after methods (which accept a `$params` array too) to the param provider, so you couildn't reuse them.

Contrary to PHPUnit I think it's improtant to have access to the parameters in the before/after methods as you'll often want to f.e. create X number of things to test against.

> I do not have access to the input parameters during this warmup phase

You should do: https://github.com/phpbench/phpbench/blob/8b769d7fe6a5f7469b66b06cf01c95867b95f309/lib/Executor/Benchmark/template/remote.template#L42

> Just to make sure I am not benchmarking something silly

Yes, "assertions" is a missing concern. At the moment sanity checking with `if ($exitCode !== 0)` then throwing an exception is probably fine unless you really care about the time to evalaute an `if` statement.

One option is to allow returning a value from the benchmark, and having a way to run assertions or whatever on the result.

 > The point above makes me wonder if it would not be better to allow an assertion API 

I do quite like the idea :slightly_smiling_face: PHPBench is an old project, with lots of legacy and stuff I'd love to change at some point, it could be _far better_ (and simpler) than it is and the current API introduces constraints which are not ideal.

> There is no tearDown counterpart to `Warmup()`

You can use `AfterMethod` ?

**Reply by theofidry** [reply]
Thanks for the detailed answers! ❤️ 

>I do quite like the idea 🙂 PHPBench is an old project, with lots of legacy and stuff I'd love to change at some point, it could be far better (and simpler) than it is and the current API introduces constraints which are not ideal.

Happy to discuss more about it anytime. I'm afraid however I would only be able to help from the backseat, unfortunately already too many OSS projects at hand 😢 

But regardless, the tool is great. I think I struggled a bit with the doc as there is really a lot to unpack there, but I admit I might have went through it a bit too fast too 😛

**Reply by dantleech** [reply]
thanks! maybe at some point I'll refocus on PHPBench but I'm distracted by other stuff currently, really appreciate any ideas and feedback as you've given here!
